### PR TITLE
Fix HTLCs rexmit out of order

### PR DESCRIPTION
**Pull request by rustyrussell**
As noted in this morning's spec call, I tracked it down.

### CODE REVIEW

File: channeld/full_channel.c [hunk 1]
```diff
@@ -758,6 +758,13 @@ enum channel_add_err channel_add_htlc(struct channel *channel,
 	else
 		state = RCVD_ADD_HTLC;
 
+	/* BOLT #2:
+	 * - MUST increase the value of `id` by 1 for each successive offer.
+	 */
+	/* This is a weak (bit cheap) check: */
+	if (htlc_get(channel->htlcs, id+1, sender))
+		status_broken("Peer sent out-of-order HTLC ids (is that you, old c-lightning node?");
```
Comment (cdecker) [inline channeld/full_channel.c:766]:
```suggestion
		status_broken("Peer sent out-of-order HTLC ids (is that you, old c-lightning node?)");
```

File: tests/test_connection.py [hunk 2]
```diff
@@ -2661,3 +2661,39 @@ def test_connection_timeout(node_factory):
...
+def test_htlc_retransmit_order(node_factory, executor):
+    NUM_HTLCS=10
+    l1, l2 = node_factory.line_graph(2,
+                                    opts=[{'may_reconnect': True,
+                                           'feerates': (7500, 7500, 7500, 7500),
+                                           'disconnect': ['=WIRE_UPDATE_ADD_HTLC-nocommit',
+                                                          '=WIRE_UPDATE_ADD_HTLC*' + str(NUM_HTLCS-1),
+                                                          '-WIRE_COMMITMENT_SIGNED']},
+                                          {'may_reconnect': True}])
+    preimages = [l2.rpc.invoice(1000, str(x), str(x))['payment_hash'] for x in range(NUM_HTLCS)]
```
Comment (cdecker) [inline tests/test_connection.py:2675]:
```suggestion
    payment_hashes = [l2.rpc.invoice(1000, str(x), str(x))['payment_hash'] for x in range(NUM_HTLCS)]
```

File: tests/test_connection.py [hunk 3]
```diff
@@ -2661,3 +2661,39 @@ def test_connection_timeout(node_factory):
...
+
+    routestep = {
+        'msatoshi': 1000,
+        'id': l2.info['id'],
+        'delay': 5,
+        'channel': '1x1x1'  # note: can be bogus for 1-hop direct payments
+    }
+    for p in preimages:
```
Comment (cdecker) [inline tests/test_connection.py:2683]:
```suggestion
    for p in payment_hashes:
```

File: tests/test_connection.py [hunk 4]
```diff
@@ -2661,3 +2661,39 @@ def test_connection_timeout(node_factory):
...
+        executor.submit(l1.rpc.sendpay, [routestep], p)
+
+    l1.daemon.wait_for_logs(['dev_disconnect'] * 2)
+    l1.rpc.call('dev-reenable-commit', [l2.info['id']])
+    l1.daemon.wait_for_log('dev_disconnect')
+
+    # Now reconnect.
+    l1.rpc.connect(l2.info['id'], 'localhost', port=l2.port)
+
+    for p in preimages:
```
Comment (cdecker) [inline tests/test_connection.py:2693]:
```suggestion
    for p in payment_hashes:
```

**Review by cdecker — APPROVED** [review]
ACK d5243d47134faaf892a423817ed11a91b10bf712

**Comment by rustyrussell** [comment]
OK, wove @cdecker fixes into commits, so self-ack:

Ack https://github.com/ElementsProject/lightning/commit/f417840779ddea4441870dd6efa9916ec0091715

**Pull request merged**